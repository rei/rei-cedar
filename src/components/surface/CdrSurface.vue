<script setup lang="ts">
import { useCssModule, computed } from 'vue';
import mapClasses from '../../utils/mapClasses';
import type { SurfaceProps, InlineCss } from '../../types/interfaces';

/** Base component for all other components */

defineOptions({ name: 'CdrSurface' });

export interface CdrSurfaceProps extends SurfaceProps {}

const props = withDefaults(defineProps<CdrSurfaceProps>(), {
  tag: 'div',
});

const style = useCssModule();
const baseClass = 'cdr-surface';
const borderClass = `${baseClass}--border`;
const radiusClass = `${baseClass}--radius`;
const shadowClass = `${baseClass}--shadow`;
const backgroundClass = `${baseClass}--background`;
const elevationClass = `${baseClass}--elevation`;
const paletteClass = `${baseClass}--palette`;

// Map of margins and padding
const marginsAndPaddingMap = {
  m: 'margin',
  mx: 'marginInline',
  my: 'marginBlock',
  ml: 'marginInlineStart',
  mr: 'marginInlineEnd',
  mt: 'marginTop',
  mb: 'marginBottom',
  p: 'padding',
  px: 'paddingInline',
  py: 'paddingBlock',
  pl: 'paddingInlineStart',
  pr: 'paddingInlineEnd',
  pt: 'paddingTop',
  pb: 'paddingBottom',
};
const marginAndPaddingKeys = Object.keys(marginsAndPaddingMap);

// Manages the props passed along to CdrSurface
const computedProps = computed(() => {
  const classes = [];
  const inlineStyles: InlineCss = {};

  // Add display
  if (props.display) {
    inlineStyles['display'] = props.display;
  }

  // Add palette
  if (props.palette) {
    classes.push(`${paletteClass}-${props.palette}`);
  }

  // Add border
  if (props.withBorder) {
    if (props.borderWidth) {
      inlineStyles['--cdr-border-width-default'] = `var(--cdr-space-inset-${props.borderWidth})`;
    }

    if (props.borderColor) {
      inlineStyles['--cdr-border-color-default'] = `var(--cdr-color-border-${props.borderColor})`;
    }

    classes.push(borderClass);
  }

  // Add radius
  if (props.radius) {
    inlineStyles['--cdr-radius-default'] = `var(--cdr-radius-${props.radius})`;
    classes.push(radiusClass);
  }

  // Add shadow
  if (props.shadow) {
    inlineStyles['--cdr-shadow-default'] = `var(--cdr-prominence-${props.shadow})`;
    classes.push(shadowClass);
  }

  // Add background
  if (props.background) {
    inlineStyles['--cdr-background-default'] = `var(--cdr-color-background-${props.background})`;
    classes.push(backgroundClass);
  }

  // Add elevation
  if (props.elevation) {
    if (props.elevation) {
      inlineStyles['--cdr-elevation-default'] = `var(--cdr-elevation-${props.elevation})`;
    }

    if (props.position) {
      inlineStyles['--cdr-position-default'] = props.position;
    }

    classes.push(elevationClass);
  }

  // Add margin and padding
  if (marginAndPaddingKeys.find((key) => props[key])) {
    marginAndPaddingKeys.forEach((key) => {
      if (props[key]) {
        const cssProperty = marginsAndPaddingMap[key];
        inlineStyles[cssProperty] = `var(--cdr-space-${props.fixed ? '' : 'scale-'}${props[key]})`;
      }
    });
  }

  return { style: inlineStyles, class: mapClasses(style, ...classes) || undefined };
});
</script>

<template>
  <component
    :is="tag"
    v-bind="computedProps"
  >
    <slot />
  </component>
</template>

<style lang="scss" module src="./styles/CdrSurface.module.scss"></style>
