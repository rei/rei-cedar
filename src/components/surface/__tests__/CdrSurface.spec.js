import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrSurface from '../CdrSurface.vue';

describe('CdrSurface', () => {
  it('default snapshot', () => {
    const wrapper = mount(CdrSurface);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('sets the default tag prop correctly', () => {
    const wrapper = mount(CdrSurface, { props: { tag: 'span' } });
    expect(wrapper.element.tagName).toBe('SPAN');
  });

  it('base classes only', () => {
    const wrapper = mount(CdrSurface, { props: { tag: 'span' } });
    expect(wrapper.classes()).toEqual(['cdr-surface']);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('shows default slot', () => {
    const wrapper = mount(CdrSurface, { slots: { default: 'Default slot' } });
    expect(wrapper.find('.cdr-surface').text()).toBe('Default slot');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies background property', () => {
    const wrapper = mount(CdrSurface, { props: { background: 'primary' } });
    expect(wrapper.attributes('data-background')).toBe('primary');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies border color property', () => {
    const wrapper = mount(CdrSurface, { props: { borderColor: 'primary' } });
    expect(wrapper.attributes('data-border-color')).toBe('primary');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies border width property', () => {
    const wrapper = mount(CdrSurface, { props: { borderWidth: 'one-x' } });
    expect(wrapper.attributes('data-border-width')).toBe('one-x');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies border style property', () => {
    const wrapper = mount(CdrSurface, { props: { borderStyle: 'solid' } });
    expect(wrapper.attributes('data-border-style')).toBe('solid');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies radius property', () => {
    const wrapper = mount(CdrSurface, { props: { borderRadius: 'soft' } });
    expect(wrapper.attributes('data-border-radius')).toBe('soft');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies box shadow property', () => {
    const wrapper = mount(CdrSurface, { props: { boxShadow: 'raised' } });
    expect(wrapper.attributes('data-box-shadow')).toBe('raised');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies multiple properties', () => {
    const wrapper = mount(CdrSurface, {
      props: {
        background: 'secondary',
        borderColor: 'primary',
        borderWidth: 'one-x',
        borderStyle: 'solid',
        borderRadius: 'soft',
        boxShadow: 'raised'
      }
    });
    expect(wrapper.attributes('data-background')).toBe('secondary');
    expect(wrapper.attributes('data-border-color')).toBe('primary');
    expect(wrapper.attributes('data-border-width')).toBe('one-x');
    expect(wrapper.attributes('data-border-style')).toBe('solid');
    expect(wrapper.attributes('data-border-radius')).toBe('soft');
    expect(wrapper.attributes('data-box-shadow')).toBe('raised');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies state-specific properties', () => {
    const wrapper = mount(CdrSurface, {
      props: {
        background: {
          rest: 'primary',
          hover: 'secondary'
        }
      }
    });
    expect(wrapper.attributes('data-background')).toBe('primary');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies loading state', () => {
    const wrapper = mount(CdrSurface, { 
      attrs: { 'data-loading': 'true' }
    });
    expect(wrapper.attributes('data-loading')).toBe('true');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies palette property', () => {
    const wrapper = mount(CdrSurface, { 
      props: { 
        palette: 'membership-subtle',
        background: 'primary'
      } 
    });
    expect(wrapper.attributes('data-palette')).toBe('membership-subtle');
    expect(wrapper.attributes('data-background')).toBe('primary');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies background image properties via style attribute', () => {
    const wrapper = mount(CdrSurface, {
      props: {
        backgroundImage: 'path/to/image.jpg',
        backgroundSize: 'contain',
        backgroundPosition: 'top left',
      }
    });
    const style = wrapper.attributes('style');
    expect(style).toContain('--cdr-surface-background-image: url(path/to/image.jpg);');
    expect(style).toContain('--cdr-surface-background-size: contain;');
    expect(style).toContain('--cdr-surface-background-position: top left;');
    expect(wrapper.element).toMatchSnapshot();
  });

  it('applies background image with default size and position', () => {
    const wrapper = mount(CdrSurface, {
      props: {
        backgroundImage: 'path/to/another.png',
      }
    });
    const style = wrapper.attributes('style');
    expect(style).toContain('--cdr-surface-background-image: url(path/to/another.png);');
    expect(style).toContain('--cdr-surface-background-size: cover;'); // Default
    expect(style).toContain('--cdr-surface-background-position: center;'); // Default
    expect(wrapper.element).toMatchSnapshot();
  });

  it('does not apply background image styles if prop is not provided', () => {
    const wrapper = mount(CdrSurface);
    const style = wrapper.attributes('style');
    // Expect style attribute to be undefined or not contain the background image variables
    expect(style === undefined || !style.includes('--cdr-surface-background-image')).toBe(true);
    expect(wrapper.element).toMatchSnapshot(); // Should match default snapshot
  });
});