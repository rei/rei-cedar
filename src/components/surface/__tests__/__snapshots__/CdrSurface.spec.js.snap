// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CdrSurface > applies background property 1`] = `
<div
  class="cdr-surface"
  data-background="primary"
>
  
  
</div>
`;

exports[`CdrSurface > applies border color property 1`] = `
<div
  class="cdr-surface"
  data-border-color="primary"
>
  
  
</div>
`;

exports[`CdrSurface > applies border style property 1`] = `
<div
  class="cdr-surface"
  data-border-style="solid"
>
  
  
</div>
`;

exports[`CdrSurface > applies border width property 1`] = `
<div
  class="cdr-surface"
  data-border-width="one-x"
>
  
  
</div>
`;

exports[`CdrSurface > applies box shadow property 1`] = `
<div
  class="cdr-surface"
  data-box-shadow="raised"
>
  
  
</div>
`;

exports[`CdrSurface > applies loading state 1`] = `
<div
  class="cdr-surface"
  data-loading="true"
>
  
  
</div>
`;

exports[`CdrSurface > applies multiple properties 1`] = `
<div
  class="cdr-surface"
  data-background="secondary"
  data-border-color="primary"
  data-border-style="solid"
  data-border-width="one-x"
  data-box-shadow="raised"
  data-radius="soft"
>
  
  
</div>
`;

exports[`CdrSurface > applies palette property 1`] = `
<div
  class="cdr-surface"
  data-background="primary"
  data-palette="membership-subtle"
>
  
  
</div>
`;

exports[`CdrSurface > applies radius property 1`] = `
<div
  class="cdr-surface"
  data-radius="soft"
>
  
  
</div>
`;

exports[`CdrSurface > applies state-specific properties 1`] = `
<div
  class="cdr-surface"
  data-background="primary"
  data-background-hover="secondary"
>
  
  
</div>
`;

exports[`CdrSurface > base classes only 1`] = `
<span
  class="cdr-surface"
>
  
  
</span>
`;

exports[`CdrSurface > default snapshot 1`] = `
<div
  class="cdr-surface"
>
  
  
</div>
`;

exports[`CdrSurface > shows default slot 1`] = `
<div
  class="cdr-surface"
>
  
  Default slot
  
</div>
`;
